### PR TITLE
Introduce ProjectCode class

### PR DESCRIPTION
**Pull request by sven1103**
Provides a new domain value object ProjectCode, that enables
to create random codes, exclude blacklisted expressions and can
parse valid codes from String representations.

### CODE REVIEW

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/ProjectCode.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+  private final String value;
+
+  private static final int LENGTH = 5;
+
+  private static final String PREFIX = "Q";
+
+  /**
+   * Creates a random project code containing of letters from the English alphabet and natural
+   * numbers.
```
Comment (jenniferboedker) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/ProjectCode.java:28]: and numbers? I thought its just letters for the Project Code. Numbers are included for e.g sample codes

Comment (sven1103) [reply projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/ProjectCode.java:28]: Actually it is numbers and letters :)

**Review by jenniferboedker — COMMENTED** [review]
Some comments to your code :)

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/ProjectCode.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    return new ProjectCode(str);
+  }
+
+  private static boolean isValid(String code) {
+    return code.startsWith(PREFIX) && (code.length() == LENGTH);
+  }
+
+  private static boolean isInvalid(String code) {
+    return !isValid(code);
+  }
```
[the remark this answers is not in the document]
Comment (sven1103) [reply projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/ProjectCode.java:128]: Readability. Negations are hard to read in conditional expressions.

File: projectmanagement/src/test/java/life/qbic/projectmanagement/domain/project/ProjectCodeSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        then:
+        thrown(IllegalArgumentException)
+
+        where:
+        blacklistedExpression << [ProjectCode.BLACKLIST]
+    }
+
+    def "Parsing a project code with a valid expression returns its object oriented form"() {
+        when:
+        ProjectCode code = ProjectCode.parse("Q" + "ABCD")
```
Comment (jenniferboedker) [inline projectmanagement/src/test/java/life/qbic/projectmanagement/domain/project/ProjectCodeSpec.groovy:33]:
```suggestion
        ProjectCode code = ProjectCode.parse("QABCD")
```

File: projectmanagement/src/test/java/life/qbic/projectmanagement/domain/project/ProjectCodeSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    def "Parsing a project code with a invalid character throws an IllegalArgumentException"() {
+        when:
+        ProjectCode.parse("Q" + wrongLength)
+
+        then:
+        thrown(IllegalArgumentException)
+
+        where:
+        wrongLength << ["ABCÖ", "YABC", "~TES"]
```
Comment (jenniferboedker) [inline projectmanagement/src/test/java/life/qbic/projectmanagement/domain/project/ProjectCodeSpec.groovy:58]:
```suggestion
        invalidChar << ["ABCÖ", "YABC", "~TES"]
```

**Review by jenniferboedker — DISMISSED** [review]

**Review by jenniferboedker — APPROVED** [review]

**Pull request merged**